File: native/Sources/NeoCoreModule.swift
import Foundation
import React

/// Identity operations, in-process through the Rust core (neo-ffi via UniFFI).
///
/// The identity file written here is byte-compatible with the CLI's
/// `neo identity generate` output (raw `NodeIdentity::to_bytes()`), so the
/// daemon started by `NeoDaemon` can consume it with `--identity`.
@objc(NeoCore)
class NeoCore: NSObject {
  @objc static func requiresMainQueueSetup() -> Bool { false }

  static var dataDir: URL {
    let base = FileManager.default.urls(for: .applicationSupportDirectory, in: .userDomainMask)[0]
    return base.appendingPathComponent("neo", isDirectory: true)
  }

  static var identityURL: URL { dataDir.appendingPathComponent("identity.key") }

  @objc func constantsToExport() -> [AnyHashable: Any]! {
    return [
      "dataDir": NeoCore.dataDir.path,
      "identityPath": NeoCore.identityURL.path,
    ]
  }

  /// Load the identity at the default path, generating one on first launch.
  /// Resolves to `{ nodeId, path, created }`.
  @objc(ensureIdentity:rejecter:)
  func ensureIdentity(_ resolve: @escaping RCTPromiseResolveBlock,
                      rejecter reject: @escaping RCTPromiseRejectBlock) {
    DispatchQueue.global(qos: .userInitiated).async {
      do {
        let fm = FileManager.default
        try fm.createDirectory(at: NeoCore.dataDir, withIntermediateDirectories: true)
        let url = NeoCore.identityURL

        var created = false
        var secret: Data
        if fm.fileExists(atPath: url.path) {
          secret = try Data(contentsOf: url)
        } else {
          secret = generateIdentity()
          guard !secret.isEmpty else {
            reject("E_RNG", "the OS RNG was unavailable while generating an identity", nil)
            return
          }
          try secret.write(to: url, options: .atomic)
          try fm.setAttributes([.posixPermissions: 0o600], ofItemAtPath: url.path)
          created = true
        }

        guard let nodeId = identityNodeId(secret: secret) else {
          reject("E_IDENTITY", "\(url.path) does not contain a valid neo identity", nil)
          return
        }
        resolve(["nodeId": nodeId, "path": url.path, "created": created])
      } catch {
        reject("E_IDENTITY", "identity setup failed: \(error.localizedDescription)", error)
      }
    }
  }

  /// The identity secret, base64-encoded, for handing to the VPN tunnel config.
  /// Kept separate from `ensureIdentity` so the raw secret is only moved when a
  /// caller explicitly needs it (e.g. to start the tunnel).
  @objc(identitySecretBase64:rejecter:)
  func identitySecretBase64(_ resolve: @escaping RCTPromiseResolveBlock,
                            rejecter reject: @escaping RCTPromiseRejectBlock) {
    DispatchQueue.global(qos: .userInitiated).async {
      let url = NeoCore.identityURL
      guard let secret = try? Data(contentsOf: url) else {
        reject("E_IDENTITY", "no identity at \(url.path) — call ensureIdentity first", nil)
        return
      }
      resolve(secret.base64EncodedString())
    }
  }
}

File: android/app/src/main/java/com/neomac/NeoCoreModule.kt
package com.neomac

import android.util.Base64
import com.facebook.react.bridge.Arguments
import com.facebook.react.bridge.Promise
import com.facebook.react.bridge.ReactApplicationContext
import com.facebook.react.bridge.ReactContextBaseJavaModule
import com.facebook.react.bridge.ReactMethod
import java.io.File
import uniffi.neo_ffi.generateIdentity
import uniffi.neo_ffi.identityNodeId

/**
 * Identity operations, in-process through the Rust core (neo-ffi via UniFFI).
 * The key file is the same raw format the `neo` CLI and the macOS app use.
 */
class NeoCoreModule(reactContext: ReactApplicationContext) :
    ReactContextBaseJavaModule(reactContext) {

  override fun getName() = "NeoCore"

  private fun identityFile(): File = File(reactApplicationContext.filesDir, "neo/identity.key")

  @ReactMethod
  fun ensureIdentity(promise: Promise) {
    Thread {
      try {
        val file = identityFile()
        file.parentFile?.mkdirs()
        var created = false
        val secret: ByteArray =
            if (file.exists()) {
              file.readBytes()
            } else {
              val fresh = generateIdentity()
              if (fresh.isEmpty()) {
                promise.reject("E_RNG", "the OS RNG was unavailable while generating an identity")
                return@Thread
              }
              file.writeBytes(fresh)
              created = true
              fresh
            }
        val nodeId = identityNodeId(secret)
        if (nodeId == null) {
          promise.reject("E_IDENTITY", "${file.path} does not contain a valid neo identity")
          return@Thread
        }
        val map = Arguments.createMap()
        map.putString("nodeId", nodeId)
        map.putString("path", file.absolutePath)
        map.putBoolean("created", created)
        promise.resolve(map)
      } catch (e: Exception) {
        promise.reject("E_IDENTITY", "identity setup failed: ${e.message}", e)
      }
    }
        .start()
  }

  /** The identity secret, base64-encoded, for handing to the VPN tunnel config. */
  @ReactMethod
  fun identitySecretBase64(promise: Promise) {
    Thread {
      try {
        val file = identityFile()
        if (!file.exists()) {
          promise.reject("E_IDENTITY", "no identity at ${file.path} — call ensureIdentity first")
          return@Thread
        }
        promise.resolve(Base64.encodeToString(file.readBytes(), Base64.NO_WRAP))
      } catch (e: Exception) {
        promise.reject("E_IDENTITY", e.message, e)
      }
    }
        .start()
  }
}
